File: src/zql/commands/set.zig
const std = @import("std");
const context = @import("../ctx.zig");
const types = @import("../../types.zig");
const collection = @import("../../collection.zig");
const mem = @import("memory");

pub fn set(ctx: *context.ZqlContext, allocator: std.mem.Allocator) anyerror!void {
    if (ctx.get_arg_count() != 2) {
        ctx.err = try std.fmt.allocPrint(allocator, "Expected 2 arguments, got {}!", .{ctx.get_arg_count()});
        return;
    }

    var arg1 = ctx.get_arg(0).?;
    var arg2 = ctx.get_arg(1).?;

    const user = ctx.get_user();
    var perm = try ctx.db.get_perm(user, "$", allocator);

    if (!perm.?.can_write) {
        ctx.err = try std.fmt.allocPrint(allocator, "Permissions denied, unable to write!", .{});
        return;
    }

    var split = std.mem.splitScalar(u8, arg1.value.String, '.');
    var current = ctx.db;

    while (split.next()) |s| {
        if (split.peek() == null) {
            try current.add(s, arg2.move(), allocator);
            break;
        }

        if (current.get(s)) |cur| {
            var tmp_perm = try current.get_perm(user, s, allocator);

            if (tmp_perm) |p| {
                perm = p;
            }

            if (!perm.?.can_write) {
                ctx.err = try std.fmt.allocPrint(allocator, "Permissions denied, unable to write!", .{});
                return;
            }

            switch (cur.value) {
                types.Value.Collection => {},
                else => {
                    ctx.err = try std.fmt.allocPrint(allocator, "Key \"{s}\" was expected to be Collection, but got {s}!", .{s, @tagName(cur.value)});
                    return;
                }
            }

            current = &cur.value.Collection;
            continue;
        }

        var v = try mem.AutoPtr(types.Value).init(allocator, .{ .Collection = try collection.Collection.init(allocator) });
        defer v.deinit();
        v.set_on_free(types.dispose);
        try current.add(s, v.move(), allocator);
        current = &v.value.Collection;
    }
}

File: src/zql/commands/set.cpp
#include <cstdio>
#include <string>

#include <types.hpp>
#include <zql/ctx.hpp>

using ZeonDB::Types::Type;
using ZeonDB::Types::Value;

void set_casual(ZeonDB::ZQL::Context* ctx) {
	auto key = ctx->get_arg(0);
	auto value = ctx->get_arg(1);

	std::string user = ctx->get_user();
	auto perms = ctx->get_perm("$");

    auto* acc = ctx->get_account(user);
    if (acc && acc->special) {
        ctx->error = "Dummy can only create account!";
        return;
    }

	if (!perms.can_write) {
		ctx->error = "Permissions deniend, unable to write!";
		return;
	}

	auto current = ctx->get_db();
	std::string s = "";
	while ((s = key->v.s.next(".")).length() > 0) {
		if (!key->v.s.peek(".")) {
			if (current->v.c.has_perms(user, s)) {
				perms = current->v.c.get_perms(user, s);
			}

			if (!perms.can_write) {
				ctx->error = "Permissions deniend, unable to write!";
				return;
			}

			current->v.c.add(s, value);
			break;
		}

		auto val = current->v.c.get(s);
		if (val != nullptr) {
			if (current->v.c.has_perms(user, s)) {
				perms = current->v.c.get_perms(user, s);
			}

			if (!perms.can_write) {
				ctx->error = "Permissions deniend, unable to write!";
				return;
			}

			val->t = Type::Collection;

			current = val;
			continue;
		}
		auto v = Value::new_collection();
		current->v.c.add(s, v);
		current = v;
	}
}

void set(ZeonDB::ZQL::Context* ctx) {
	size_t arg_count = ctx->arg_count();
	switch (arg_count) {
		case 2:
			set_casual(ctx);
			break;
		default:
			ctx->error = R"(help: set (local) <key> (<value>)(from (local) <key2>)
set <key> <value> -- set <value> at specific <key>

key syntax: key@branch[index])";
	}
}
